File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081334Z_claude_claude-sonnet-5_php-symfony/src/Downtime/DowntimeController.php
<?php

namespace App\Downtime;

use App\Storage\Database;
use App\Support\Json;
use App\Support\Validators;
use PDO;
use Symfony\Component\HttpFoundation\JsonResponse;

/** Handlers for downtime crafting endpoints (/v1/campaigns/{id}/downtime/crafting/*). */
final class DowntimeController
{
    public function createProject(array $body, string $campaignId): JsonResponse
    {
        $id = $body['id'] ?? null;
        $characterId = $body['character_id'] ?? null;
        $itemSlug = $body['item_slug'] ?? null;
        $daysRequired = $body['days_required'] ?? null;
        $costGp = $body['cost_gp'] ?? null;

        if (!Validators::isValidId($id) || !Validators::isValidId($characterId)
            || !Validators::isValidSlug($itemSlug)
            || !Validators::isValidInt($daysRequired) || (int) $daysRequired <= 0
            || !Validators::isValidInt($costGp) || (int) $costGp < 0) {
            return Json::error('invalid request');
        }
        $daysRequired = (int) $daysRequired;
        $costGp = (int) $costGp;

        $db = Database::connection();
        $stmt = $db->prepare('SELECT id FROM campaigns WHERE id = ?');
        $stmt->execute([$campaignId]);
        if ($stmt->fetchColumn() === false) {
            return Json::error('campaign not found', 404);
        }

        $stmt = $db->prepare('SELECT id FROM campaign_characters WHERE id = ? AND campaign_id = ?');
        $stmt->execute([$characterId, $campaignId]);
        if ($stmt->fetchColumn() === false) {
            return Json::error('character not found', 404);
        }

        $stmt = $db->prepare('SELECT id FROM campaign_crafting WHERE id = ?');
        $stmt->execute([$id]);
        if ($stmt->fetchColumn() !== false) {
            return Json::error('crafting project already exists', 409);
        }

        $insert = $db->prepare(
            'INSERT INTO campaign_crafting (id, campaign_id, character_id, item_slug, days_required, cost_gp, days_completed, status)
             VALUES (?, ?, ?, ?, ?, ?, 0, ?)'
        );
        $insert->execute([$id, $campaignId, $characterId, $itemSlug, $daysRequired, $costGp, 'active']);

        return new JsonResponse([
            'id' => $id,
            'character_id' => $characterId,
            'item_slug' => $itemSlug,
            'days_required' => $daysRequired,
            'days_completed' => 0,
            'status' => 'active',
        ], 201);
    }

    public function advance(array $body, string $campaignId, string $projectId): JsonResponse
    {
        $days = $body['days'] ?? null;
        if (!Validators::isValidInt($days) || (int) $days <= 0) {
            return Json::error('invalid request');
        }
        $days = (int) $days;

        $db = Database::connection();
        $stmt = $db->prepare('SELECT id, item_slug, days_required, days_completed, status FROM campaign_crafting WHERE id = ? AND campaign_id = ?');
        $stmt->execute([$projectId, $campaignId]);
        $project = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($project === false) {
            return Json::error('crafting project not found', 404);
        }

        if ($project['status'] === 'complete') {
            return Json::error('crafting project already complete', 409);
        }

        $daysRequired = (int) $project['days_required'];
        $daysCompleted = min($daysRequired, (int) $project['days_completed'] + $days);
        $status = $daysCompleted >= $daysRequired ? 'complete' : 'active';

        $update = $db->prepare('UPDATE campaign_crafting SET days_completed = ?, status = ? WHERE id = ?');
        $update->execute([$daysCompleted, $status, $projectId]);

        if ($status === 'complete') {
            $itemSlug = $project['item_slug'];
            $stmt = $db->prepare("SELECT quantity FROM campaign_inventory WHERE campaign_id = ? AND item_slug = ? AND owner = 'party'");
            $stmt->execute([$campaignId, $itemSlug]);
            $existing = $stmt->fetchColumn();

            if ($existing === false) {
                $insert = $db->prepare("INSERT INTO campaign_inventory (campaign_id, item_slug, owner, quantity) VALUES (?, ?, 'party', 1)");
                $insert->execute([$campaignId, $itemSlug]);
            } else {
                $updateInv = $db->prepare("UPDATE campaign_inventory SET quantity = ? WHERE campaign_id = ? AND item_slug = ? AND owner = 'party'");
                $updateInv->execute([(int) $existing + 1, $campaignId, $itemSlug]);
            }
        }

        return new JsonResponse([
            'id' => $project['id'],
            'days_completed' => $daysCompleted,
            'status' => $status,
        ]);
    }
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051117Z_claude_claude-sonnet-5_ruby-rails/app/controllers/downtime_controller.rb
# Downtime crafting projects: creation and day-by-day progress tracking.
# Projects are stored inline on the campaign hash under :crafting_projects,
# keyed by project id, and persisted through CAMPAIGNS.persist.
class DowntimeController < ApplicationController
  def create_crafting_project
    campaign = CAMPAIGNS[params[:campaign_id]]
    if campaign.nil?
      render json: { error: 'campaign not found' }, status: :not_found
      return
    end

    id = params[:id]
    character_id = params[:character_id]
    item_slug = params[:item_slug]
    days_required = params[:days_required]
    cost_gp = params[:cost_gp]

    unless id.is_a?(String) && !id.empty?
      render json: { error: 'invalid id' }, status: :bad_request
      return
    end

    unless character_id.is_a?(String) && !character_id.empty?
      render json: { error: 'invalid character_id' }, status: :bad_request
      return
    end

    unless item_slug.is_a?(String) && !item_slug.empty?
      render json: { error: 'invalid item_slug' }, status: :bad_request
      return
    end

    unless valid_integer?(days_required) && days_required.to_i.positive?
      render json: { error: 'invalid days_required' }, status: :bad_request
      return
    end

    unless valid_integer?(cost_gp) && cost_gp.to_i >= 0
      render json: { error: 'invalid cost_gp' }, status: :bad_request
      return
    end

    campaign[:crafting_projects] ||= {}
    if campaign[:crafting_projects].key?(id)
      render json: { error: 'duplicate id' }, status: :conflict
      return
    end

    campaign[:crafting_projects][id] = {
      character_id: character_id,
      item_slug: item_slug,
      days_required: days_required.to_i,
      days_completed: 0,
      cost_gp: cost_gp.to_i,
      status: 'active'
    }
    CAMPAIGNS.persist(params[:campaign_id])

    render json: crafting_payload(id, campaign[:crafting_projects][id], full: true), status: :created
  end

  def advance_crafting_project
    campaign = CAMPAIGNS[params[:campaign_id]]
    if campaign.nil?
      render json: { error: 'campaign not found' }, status: :not_found
      return
    end

    project = campaign[:crafting_projects] && campaign[:crafting_projects][params[:project_id]]
    if project.nil?
      render json: { error: 'crafting project not found' }, status: :not_found
      return
    end

    days = params[:days]
    unless valid_integer?(days) && days.to_i.positive?
      render json: { error: 'invalid days' }, status: :bad_request
      return
    end

    if project[:status] == 'complete'
      render json: { error: 'crafting project already complete' }, status: :conflict
      return
    end

    project[:days_completed] = [project[:days_completed] + days.to_i, project[:days_required]].min

    if project[:days_completed] >= project[:days_required]
      project[:status] = 'complete'
    end

    CAMPAIGNS.persist(params[:campaign_id])

    render json: crafting_payload(params[:project_id], project)
  end

  private

  def crafting_payload(id, project, full: false)
    payload = { id: id }
    if full
      payload[:character_id] = project[:character_id]
      payload[:item_slug] = project[:item_slug]
      payload[:days_required] = project[:days_required]
    end
    payload[:days_completed] = project[:days_completed]
    payload[:status] = project[:status]
    payload
  end
end
